vendors/mikrotik/generate_main.go

```diff
@@ -0,0 +1,110 @@
+//+build ignore
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"layeh.com/radius/dictionary"
+	"layeh.com/radius/dictionarygen"
+
+	"golang.org/x/net/html"
+)
+
+func main() {
+	resp, err := http.Get(`https://wiki.mikrotik.com/wiki/Manual:RADIUS_Client/vendor_dictionary`)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer resp.Body.Close()
+
+	// the dictionary is inside a <pre> tag, so... tokenize it out
+	var extractedBody []byte
+	var state int
+
+	tokenizer := html.NewTokenizer(resp.Body)
+	for {
+		tokenType := tokenizer.Next()
+		if tokenType == html.ErrorToken {
+			err := tokenizer.Err()
+			if err == io.EOF {
+				break
+			} else {
+				log.Fatal(err)
+			}
+		}
+
+		switch state {
+		case 0:
+			if tokenType == html.StartTagToken {
+				token := tokenizer.Token()
+				if token.Data == "pre" {
+					state = 1
+				}
+			}
+
+		case 1:
+			if tokenType == html.TextToken {
+				extractedBody = append(extractedBody, []byte(tokenizer.Token().Data)...)
+			} else {
+				state = 2
+			}
+		}
+	}
+	if len(extractedBody) < 500 {
```

Comment:
Why this test to see if the body was extracted? Can't we just check the result of the parser?

Reply:
I added this on the assumption that the parser is brittle to layout changes in the upstream page. A length check catches the case where the tokenizer didn't find anything, which could be done by checking state, and also catches the case where it picked out the contents of an unrelated `<pre>` tag or where the `<pre>` tag no longer directly contains the text of the dictionary.

Despite the ugliness here, this page appears to be the official source of this data. I thought about using the FreeRADIUS copy of this dictionary to skip this HTML processing, but that copy isn't _quite_ equivalent today and in the recent past it was six years out of date. Parsing HTML this way is brittle but in exchange it will a) track the official vendor dictionary or b) fail completely; eliminating the telephone game with third party sources seemed like a worthwhile tradeoff.

Reply:
Sorry, I guess I wasn't totally clear.

Rather than have the length as a success test. rather check the result of the dictionary parser (e.g. ensure that at least 1 attribute was extracted).

Reply:
Or test that the vendor "Mikrotik" was extracted.

Reply:
Ah, good point. I now check that the dictionary contains a vendor with the right number, and that the vendor contains at least as many attributes as it does now.